core/block_validator.go

```diff
@@ -68,14 +68,21 @@ func (v *BlockValidator) ValidateBody(block *types.Block) error {
 	if hash := types.DeriveSha(block.Transactions(), trie.NewStackTrie(nil)); hash != header.TxHash {
 		return fmt.Errorf("transaction root hash mismatch (header value %x, calculated %x)", header.TxHash, hash)
 	}
-	// Withdrawals are present after the Shanghai fork.
-	if header.WithdrawalsHash != nil {
-		// Withdrawals list must be present in body after Shanghai.
-		if block.Withdrawals() == nil {
-			return fmt.Errorf("missing withdrawals in block body")
-		}
-		if hash := types.DeriveSha(block.Withdrawals(), trie.NewStackTrie(nil)); hash != *header.WithdrawalsHash {
-			return fmt.Errorf("withdrawals root hash mismatch (header value %x, calculated %x)", *header.WithdrawalsHash, hash)
+	if v.config.IsShanghai(header.Time) {
+		// Withdrawals are present after the Shanghai fork.
+		if header.WithdrawalsHash != nil {
+			// Withdrawals list must be present in body after Shanghai.
+			if block.Withdrawals() == nil {
+				return fmt.Errorf("missing withdrawals in block body")
+			}
+			if hash := types.DeriveSha(block.Withdrawals(), trie.NewStackTrie(nil)); hash != *header.WithdrawalsHash {
+				return fmt.Errorf("withdrawals root hash mismatch (header value %x, calculated %x)", *header.WithdrawalsHash, hash)
+			}
+		} 
+		return fmt.Errorf("withdrawals hash is missing")
+	} else {
+		if block.Withdrawals() != nil {
```

Comment:
I think you should also check `WithdrawalsHash` here as they should not be set pre-shanghai